clang/include/clang/AST/Expr.h

```diff
@@ -7379,6 +7379,14 @@ class RecoveryExpr final : public Expr,
   friend class ASTStmtWriter;
 };
 
+/// Insertion operator for diagnostics.  This allows sending
+/// Expr into a diagnostic with <<.
+inline const StreamingDiagnostic &operator<<(const StreamingDiagnostic &DB,
+                                             const Expr *E) {
+  DB.AddTaggedVal(reinterpret_cast<uint64_t>(E), DiagnosticsEngine::ak_expr);
```

Comment:
So is this correct, to cast a pointer to `uint64_t`? It looks like we do this and in about half the other place we use `intptr_t`, which I would be much less suspicious of.

CC @AaronBallman @erichkeane

Reply:
`intptr_t` /`uintptr_t` I think avoids any 'round trip' errors by standard, but I think `uint64_t` is going to be the same type ANYWAY so this is probably harmless at least.  I think the suggestion to use `intptr_t`/`uintptr_t` is a good one that should happen.

Reply:
If we care that uint64_t might not round trip a pointer in theory, even though it should always in practice, then do we care uintptr_t is not guaranteed to be provided in theory, even though it should always be in practice?

Reply:
> If we care that uint64_t might not round trip a pointer in theory, even though it should always in practice, then do we care uintptr_t is not guaranteed to be provided in theory, even though it should always be in practice?

THAT we don't care about because : 1- it is a compile-time diagnosed missing 'thing', and 2- We know all our supported host platforms have (and presumably always will have) that type.

So it is a matter of "potentially ill-formed" vs "potential-UB".

Either way, it doesn't MATTER very much, but from a 'most correct' perspective (as frankly, `uint64_t` and `uintptr_t` are probably the same type, and I would bet good money that none of our platforms would ever change that/never give `uintptr_t` its own representation to take advantage of this UB), but just changing the cast type here is the right thing to do.

Reply:
It's not just here though, this is the same for all other kinds, and the same for the parameter type, and the same for where we store this.

It wouldn't change much regarding potential UB if we wrote uintptr_t here, but then this would implicitly cast to uint64_t anyway.

Now if we change that to uintptr_t, then that would be bad in 32 bit platforms, as this is used to pack other things besides pointers. So presumably we want to change the parameter and structure types here to something that's at least as big as both types, which would be uintmax perhaps?

Or maybe just assert somewhere that uint64 is at least as big as uintptr?

Reply:
As far as correctness: we need to smuggle pointers through a `uintptr_t`.  Beyond that, there is no harm then again casting that immediately to the `uint64_t`.

So the 'fixit' suggestion is:
```suggestion
  DB.AddTaggedVal(static_cast<uint64_t>(reinterpret_cast<uintptr_t>(E)), DiagnosticsEngine::ak_expr);
```

The underlying type here is fine, and SHOULD stay `uint64_t`.  But the `reinterpret_cast` of pointer-to-int itself is the problematic part.  If you'd like, you can put a static-assert that `sizeof(uintptr_t)` is <= `sizeof(uint64_t)`, but I'm not horribly worried about 128 bit pointers

Reply:
But presumably, I'd have to go on and do this for all diagnostics stream operators, not just this one, otherwise it would not be helpful to the next individual, who just
like me, did the same way it was done somewhere else :)

Reply:
We probably SHOULD, and you can do so without review if you care to.  Else, we can just make @shafik do this, since he started it.

Reply:
I am not super sympathetic here, `reinterpret_cast` in a code review should have both the author and the reviewer on high alert and they should carefully examine the code and if we suspect the same pattern is used elsewhere we should confirm it is used consistently. Which in this case it is not. 

The PR was turned over pretty quickly w/o giving any of the other long list of reviewers a chance to comment.

Reply:
Yes, I had carefully reviewed it and I think it was fine, given the circumstances.

For reference, here are all the other uses:
```C++
clang/include/clang/AST/Attr.h:  DB.AddTaggedVal(reinterpret_cast<uint64_t>(At), DiagnosticsEngine::ak_attr);
clang/include/clang/AST/Expr.h:  DB.AddTaggedVal(reinterpret_cast<uint64_t>(E), DiagnosticsEngine::ak_expr);
clang/include/clang/AST/NestedNameSpecifier.h:  DB.AddTaggedVal(reinterpret_cast<uint64_t>(NNS),
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(reinterpret_cast<intptr_t>(Str),
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(I, DiagnosticsEngine::ak_sint);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(I, DiagnosticsEngine::ak_sint);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(I, DiagnosticsEngine::ak_sint);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(I, DiagnosticsEngine::ak_sint);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(I, DiagnosticsEngine::ak_uint);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(I, DiagnosticsEngine::ak_uint);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(I, DiagnosticsEngine::ak_uint);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(static_cast<unsigned>(I), DiagnosticsEngine::ak_tokenkind);
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(reinterpret_cast<intptr_t>(II),
clang/include/clang/Basic/Diagnostic.h:  DB.AddTaggedVal(reinterpret_cast<intptr_t>(DC),
clang/include/clang/Basic/PartialDiagnostic.h:        DB.AddTaggedVal(DiagStorage->DiagArgumentsVal[i],
clang/include/clang/Sema/ParsedAttr.h:  DB.AddTaggedVal(reinterpret_cast<uint64_t>(At.getAttrName()),
clang/include/clang/Sema/ParsedAttr.h:  DB.AddTaggedVal(reinterpret_cast<uint64_t>(At->getAttrName()),
clang/include/clang/Sema/ParsedAttr.h:  DB.AddTaggedVal(reinterpret_cast<uint64_t>(CI.getAttrName()),
clang/include/clang/Sema/ParsedAttr.h:  DB.AddTaggedVal(reinterpret_cast<uint64_t>(CI->getAttrName()),
```

So 6 uint64_t vs 3 intptr_t (excluding this patch). Without going on and changing everything else, there is no single answer here which would be satisfy everyone.